### PR TITLE
Week 1 homework: Mon - Wed

### PR DESCRIPTION
**Pull request by jamesedwardferguson**
How difficult did you find this (out of 10)? 0 being no problems at all, 10 being impossible;
8 out of 10 - I've found the last couple of days tough going.

Was there anything that you struggled with?;
Functions and loops the most.

Is there anything that you'd like some further information on?;

Roughly how long did it take?

### CODE REVIEW

**Comment by jamesedwardferguson** [comment]
MTA homework:-

How difficult did you find this (out of 10)? 0 being no problems at all, 10 being impossible;
9
Was there anything that you struggled with?;
Even with my planning, I struggled to breakdown the problem (code) into steps and functions.

Is there anything that you'd like some further information on?;
I think I just need to keep going over the JS basics, until it finally sticks!

Roughly how long did it take?
I probably spent 10hrs +

**Comment by cjbarnaby** [comment]
Hey James, 

Good work with the homework so far. I know you feel like you're struggling a bit, but that's pretty normal - we've gone through a _lot_ of stuff this last week, and it involves a lot of concepts that can feel pretty abstract to start with. 

Feel free to grab me _any_ time you've got a question or you're struggling with something, whether it's the warmup, the homework, after a code-along in class, whatever. That's what they're paying me to do, so if you don't, it's like I'm stealing or something 😸 .

Some suggestions based on the homework submitted to date:

## General comments 

### Pseudocode 

Looking at a problem can be daunting, and it's easy to get overwhelmed by trying to think about too much at once. A good practice to get into is writing pseudocode before you write a line of code - in plain english, set out (as a series of steps) what your program needs to do, and attack one step at a time. I find it helps to break problems down into components, then break those components down into smaller steps.

### `if` / `else if` trickle-down conditions 

Re: the Grade Assigner - perfectly valid code there, but we can often take advantage of the fact that an `if` / `else if` / `else` statement will:
- not test any conditions after a condition has been satisfied; and, conversely, 
- only test a condition if all previous conditions have failed; 

We can often use that to simplify our conditions. So, for example:

```js
if (testResult >= 90) {
  console.log("Result A");
} else if (testResult >= 80 && testResult <=90) {
  console.log("Result B");
// etc
}

The second condition will _only ever_ be tested if the first condition fails/evaluates to `false`. So, if the `else if` condition there is being tested, we already _know_ that `(testResult < 90)`. So the only test the `else if` needs to check is `(testResult >= 80)`. 

### Formatting and style

Generally really good! Might seem a trivial point, but future you and other developers will thank you for having a good, consistently-applied style. There's no real universally accepted style guide for JS, but the [Idiomatic Style Guide](https://github.com/rwaldron/idiomatic.js#user-content-table-of-contents)'s style suggestions are pretty non-controversial. 

A few little suggestions:

All three examples are illustrated in the code snippet below these bullet points.

- *indenting* - generally, your code was well indented, but keep an eye on it - the first thing I do when I'm debugging someone's code is fix the indenting, just to make it easier to read and follow, and therefore easier to spot issues. 
- *whitespace* - again, generally good use of whitespace, but use it liberally (eg between operators and their operands, between parentheses and their arguments, etc) - it makes code so much easier to read.
- *semicolons* - don't forget to close your statements with semi-colons (note exceptions below, though). 99.9% of the time, a missing semicolon won't break anything, and even though browsers will add your missing semicolons in for you, it's a good habit to put them in where they're supposed to go - there are some technical reasons for this, but also, a lot of developers are semicolon Nazis. [This is a really good, quick guide on using semicolons in JS](https://www.codecademy.com/blog/78) (despite the fact that it talks about semicolons being 'optional' in certain places - they're not optional, really; statement missing a semicolon is not actually valid JS, you're just relying on the browser to put your semicolons in for you). 

Examples of the three above suggestions, from the **Pluralizer** homework:

```js
// This is all fine...
else if (noun.slice(-1)==="y"){
newWord = noun.slice(0, -1);
console.log(newWord + "ies");
}

// But this is more readable...
else if ( noun.slice( -1 ) === "y" ) {
  newWord = noun.slice( 0, -1 );
  console.log( newWord + "ies" );
};
```

### Arrays of objects 

Re: the structure of your `readingList` array in the **Reading List** homework - if we have an array of objects, we don't want to actually declare those objects as variables within the object (this syntax is invalid: `var myArray = [var myFirstObject = {},  var mySecondObject = {}]`. 

So, instead of:
```js
var ReadingList = [
   var bookOne = {
    description: "A book about stuff",
    title: " book 1",
    author: "Richard Cheese",
    alreadyRead:(true),
  };
   var bookTwo = {
    description: "Another book about stuff",
    title: " book 2",
    author: "Joe Bloggs",
    alreadyRead: (false),
  };
];
```

We would do this - note that we're just setting object literals as elements in the array (no `var` keyword): 
```js
var readingList = [
   {
    description: "A book about stuff",
    title: " book 1",
    author: "Richard Cheese",
    alreadyRead:(true),
  },
   {
    description: "Another book about stuff",
    title: " book 2",
    author: "Joe Bloggs",
    alreadyRead: (false),
  }
];
```

If we wanted to store those objects in variables, we would declare them outside of the array, and just reference them in the array. Example:

```js

var bookOne =    {
    description: "A book about stuff",
    title: " book 1",
    author: "Richard Cheese",
    alreadyRead:(true),
};
var bookTwo = {
    description: "Another book about stuff",
    title: " book 2",
    author: "Joe Bloggs",
    alreadyRead: (false),
};

var readingList = [bookeOne, bookTwo];
```

Your `for` loop iterating over the array is perfect, though! 

### Geometry Lab, Bank, Credit Card Validator, 

Let me know if you want to work through any of these

I'll post MTA feedback in a separate comment.  

Badger

**Comment by cjbarnaby** [comment]
## MTA

As Jack said, MTA is really tough. And remember, on Friday, you'd only been learning JavaScript for 4 days. 

Let me know if you want to work through this a little more, and feel free to have another crack at it and push it up to GitHub for me to review. But what you've got right now is well on the way to having it done - you're taking an input of stations and lines, you're finding the relevant array of stations, and using the index of the start and end stations in that array to get all the stops between one station and another. That's basically it. The only real complexity is in:

- multi-line trips: this is just the same as any other trip, except that you'll have Union Square as the end station of the first trip and the start station of the second trip. Think about how you can re-use the code in your current planTrip function to save yourself from re-writing code; and
- travelling backwards (ie, from one station to another station with a lower index in the station array). This should just be a matter of having a for loop that decrements, rather than increments, the value of `i`. 

I'll make one suggestion to keep you moving - I would have your `planTrip` function call other functions for you. So, a multi-line trip, or a trip going backwards needs to do basically the same thing as what you've got in your `planTrip` function right now. So I'm going to move most of that code into a new function that we can call when we need it:

```js

var getStations = function(line, startStation, endStation) {
  var startingStationIndex = startingLine.indexOf(startStation);
  var endingStationIndex = startingLine.indexOf(endStation);
  var journey = startingLine.slice(startingStationIndex, endingStationIndex);
  console.log(journey);
  console.log(journey.length);
};
var planTrip = function(startLine, startStation, endLine, endStation) {
  var startingLine = lines[startLine];
  if (startLine === endLine) {
    getStations(startLine, startStation, endStation)
  } else {
    // we'd want to call getStations for the first trip, ending at union square, and then for the second trip, starting at union square
  }
};
```

Obviously your final thing needs a lot more going on, but that's just how you might start breaking this up into smaller functions at some point. 

Badger.
